format/Stream.fbs

```diff
@@ -0,0 +1,48 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+include "Message.fbs";
+
+namespace org.apache.arrow.flatbuf;
+
+/// ----------------------------------------------------------------------
+/// Arrow stream metadata
+///
+///
+/// The format of the stream is:
+///   - StreamHeader
+///   - Block*
+///
+/// Block:
+///   1. Metadata Length (i32 little endian): Includes 4 byte meta size prefix. Size of 3).
+///   2. Body Length (i32 little endian). Size of 4).
+///   3. RecordBatch header flatbuffer (with i32 size prefix)
+///   4. Data buffers
+/// No padding between the 4 elements.
+
+table StreamHeader {
+  version: org.apache.arrow.flatbuf.MetadataVersion;
+
+  schema: org.apache.arrow.flatbuf.Schema;
+
+  /// Total number of batches in the stream. -1 to indicate unknown.
+  totalBatches: int;
+
+  /// TODO: we can add a flag/length if there are dictionaries.
```

Comment:
We haven't committed to anything yet, the only requirement is that we establish a dictionary numbering convention in stream/IPC settings (https://github.com/apache/arrow/blob/master/format/Message.fbs#L168), and in the File.fbs, the dictionary blocks are separate from the record batches https://github.com/apache/arrow/blob/master/format/File.fbs#L31

Since the presumption is that a dictionary would be shared amongst a set of record batches (compared with Parquet where the dictionary may change between row groups), it seems reasonable to put all the dictionaries first. Structurally a dictionary is a RecordBatch with a single field. So if a field in the schema has a dictionary `i`, then by convention that is the `i`-th dictionary that appears in the stream. 

So if we had the schema

```
f0: list<string: dictionary-encoded> [non-dictionary-encoded]
f1: int64
f2: binary, dictionary-encoded
```

then we assign dictionary ids 0 and 1 in the metadata and write the dictionaries as the first two record batches in the stream. It's probably flexible enough to have a `numDictionaries` count. Any use cases this might not capture?